modyn/trainer_server/internal/trainer/remote_downsamplers/irreducible_loss_producer.py

```diff
@@ -0,0 +1,104 @@
+import io
+import json
+import logging
+import pathlib
+import tempfile
+from typing import Any, Union
+
+import grpc
+import torch
+from modyn.common.ftp import download_trained_model
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+
+# pylint: disable-next=no-name-in-module
+from modyn.model_storage.internal.grpc.generated.model_storage_pb2 import FetchModelRequest, FetchModelResponse
+from modyn.model_storage.internal.grpc.generated.model_storage_pb2_grpc import ModelStorageStub
+from modyn.utils import dynamic_module_import, grpc_connection_established
+
+logger = logging.getLogger(__name__)
+
+
+class IrreducibleLossProducer:
+    def __init__(
+        self,
+        per_sample_loss: torch.nn.modules.loss,
+        modyn_config: dict,
+        rho_pipeline_id: int,
+        il_model_id: int,
+        device: str,
+    ) -> None:
+        self.model = IrreducibleLossProducer._load_il_model(modyn_config, rho_pipeline_id, il_model_id)
+        self.model.model.eval()
+        self.loss_cache: dict[int, torch.Tensor] = {}
+        self.device = device
+        self.per_sample_loss_fct = per_sample_loss
+
+    def get_irreducible_loss(
+        self, sample_ids: list[int], forward_input: Union[dict[str, torch.Tensor], torch.Tensor], target: torch.Tensor
+    ) -> torch.Tensor:
+        # use sample_ids to index into the precomputed loss values, if they are all in the cache
+        # or use forward_input to compute the loss values
+        fullly_in_cache = all(sample_id in self.loss_cache for sample_id in sample_ids)
+        if fullly_in_cache:
+            # torch.stack creates a new tensor
+            return torch.stack([self.loss_cache[sample_id] for sample_id in sample_ids]).to(self.device)
```

Comment:
I don't fully get the caching logic yet, I think. My first question is why do we need a cache? Do we ever call `get_irreducbile_loss` multiple times? Because if the loss stays on the GPU, that may eat up VRAM.

My second question is why, if the loss is not fully in cache, we recompute everything, instead of just the data where we have not cached it yet. Anyways, I think I am more confused why we need the cache, if we onlycall get_irreducible_loss in the function inform_samples once.

Reply:
Per discussion:
1. we need a cache because we (possibly) will train for multiple epochs, i.e. passing the data multiple times. We will call `get_irreducbile_loss` multiple times for the same sample. The cache does not stay always on GPU: it always stays on CPU and when it is needed, the needed part is moved to gpu.
2. The time to compute partial batch should be the same as compute the full batch. Also, we shouldn't have a partial-hit case (i.e. either fully miss or fully hit), because this transition should happen per-epoch: i.e. during first epoch every sample is missing in the cache, after the first epoch every sample is in the cache. We won't (in the most cases) have duplicate samples within an epoch.